### PR TITLE
DrivAerML: Prediction-Error-Weighted Surface Sampling (Hard Example Mining)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML samples 50,000 surface points uniformly at random per mesh. This means the model spends equal training effort on easy (flat) and hard (high-curvature, separation, stagnation) regions. As training converges to 3.833%, the remaining error is increasingly concentrated in specific hard regions (leading edges, trailing edges, wheel arches, mirrors, underbody separation).

By implementing **importance sampling based on per-point prediction error from the previous epoch**, we bias sampling toward high-error regions. This is a form of hard example mining adapted to point-cloud regression: points where the model is wrong get overrepresented in the next epoch's samples. This is fundamentally different from all other DrivAerML experiments — it changes **WHAT** the model sees rather than **HOW** it learns.

**Caveat:** With batch-size=1, each training step is a single mesh. The importance sampling happens WITHIN each mesh — we still see all 394 meshes per epoch, but we preferentially sample high-error surface points from each mesh.

## Baseline

Current best DrivAerML (PR #3072, eren/dm-ema-9995-gc05):
- `val_primary/surface_rel_l2_pct`: **3.833%** at epoch 511
- test: 4.685%
- W&B run: `ncl1dh88`
- Config: AdamW lr=5e-4, gc=0.5, no WD, 4L/512d/8H, EMA=0.9995, T_max=30, 394 batches/epoch

**External target:** <3.71% (AB-UPT). Gap: ~0.123 pp.

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in `train.py` where a local variable (around line 1646–1648) shadows the function defined around line 1428, causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` wherever it conflicts.

## Implementation Guidance

1. Add a `--error-weighted-sampling` flag (boolean, default `False`)
2. Add a `--error-sampling-temperature` flag (float, default `1.0`)
3. After the first epoch (or a short warm-up), compute per-point prediction errors during eval
4. Store error-weighted sampling probabilities per mesh (keyed by mesh ID)
5. During training, instead of uniform random sampling of 50k points, sample proportional to error weights from the **previous epoch**
6. Use temperature-scaled softmax on errors:
   ```python
   prob = softmax(errors / temperature)
   ```
   where `temperature` controls aggressiveness — lower = more focused on hard points
7. **First epoch** must use uniform sampling (no errors available yet). Optionally warm up for 5–10 epochs before switching.
8. Log the effective sample diversity (e.g., entropy of the sampling distribution) to W&B so we can monitor how concentrated sampling becomes over time.

## Trial Commands

**Trial 1 — Error-weighted sampling, temperature=1.0:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --error-weighted-sampling \
  --error-sampling-temperature 1.0 \
  --wandb_group dm-error-sampling
```

**Trial 2 — Error-weighted sampling, temperature=0.5 (more aggressive):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --error-weighted-sampling \
  --error-sampling-temperature 0.5 \
  --wandb_group dm-error-sampling
```

## Success Criteria

Report `val_primary/surface_rel_l2_pct` at best epoch for both trials.

**Target: beat 3.833%**

Also report:
- The entropy/diversity of the sampling distribution over epochs (is it converging to a small set of hard points, or staying diffuse?)
- Whether the model starts to diverge or overfit with the more aggressive temperature=0.5 setting
- Any qualitative observations about which mesh regions are being oversampled

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jet:

**Bug fix: missing `torch.no_grad()` in `compute_training_error_maps`**

Both trials (T=1.0 and T=0.5) crashed with `CUBLAS_STATUS_INTERNAL_ERROR` at epoch 5 when error map computation first triggered. Root cause: the `compute_training_error_maps` function was running ~2400 forward passes (394 cases × ~6 chunks each) without `torch.no_grad()`, causing gradient graph accumulation and OOM that manifested as a CUBLAS bf16 matmul error.

Fixed by wrapping the entire computation loop in `torch.no_grad()`. Both trials relaunched — epochs 1-4 will retrain with uniform sampling (identical to before), and epoch 5+ will now correctly compute error maps and activate error-weighted sampling.

Also fixed Python stdout buffering issue (added `PYTHONUNBUFFERED=1`) that was hiding all log output in earlier attempts.

**Comment by tcapelle** [comment]
STUDENT jet:

## Results

### Summary: Hypothesis DEAD — Error-weighted sampling causes immediate training divergence

**Baseline**: 3.833% val_primary/surface_rel_l2_pct (30 epochs)

### Trial 1: Temperature = 1.0 (linear power-law: probs ∝ errors)

| Epoch | val_primary/surface_rel_l2_pct | loss  | grad_norm_mean | Sampling |
|-------|-------------------------------|-------|----------------|----------|
| 1     | 60.19%                        | 0.928 | 9.57           | uniform  |
| 2     | 50.81%                        | 0.555 | 4.20           | uniform  |
| 3     | 45.70%                        | 0.441 | 4.39           | uniform  |
| 4     | 42.69%                        | 0.379 | 4.17           | uniform  |
| 5     | **40.79%** (best)             | 0.336 | 3.42           | uniform* |
| 6     | 57.03% ⚠️ DIVERGED           | 1.657 | 15.60          | error-weighted |

\*Error maps computed after epoch 5 eval, first used in epoch 6 training.

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/7amzkwq3
- Peak VRAM: ~6.8 GB

### Trial 2: Temperature = 0.5 (quadratic power-law: probs ∝ errors²)

| Epoch | val_primary/surface_rel_l2_pct | loss  | grad_norm_mean | Sampling |
|-------|-------------------------------|-------|----------------|----------|
| 1     | 60.55%                        | 0.930 | 9.63           | uniform  |
| 2     | 50.68%                        | 0.547 | 4.64           | uniform  |
| 3     | 44.14%                        | 0.418 | 4.42           | uniform  |
| 4     | 41.40%                        | 0.364 | 4.55           | uniform  |
| 5     | **39.31%** (best)             | 0.318 | 3.66           | uniform* |
| 6     | 77.78% ⚠️ DIVERGED           | 2.224 | 28.70          | error-weighted |

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/62zxlf2x
- Peak VRAM: ~6.8 GB

### Commands

```bash
# Trial 1
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --error-weighted-sampling --error-sampling-temperature 1.0 \
  --wandb-group dm-error-sampling --wandb-name "jet/dm-error-sampling-t1.0" --agent jet

# Trial 2 (same but --error-sampling-temperature 0.5 on GPU 1)
```

### What happened

**Error-weighted sampling causes immediate catastrophic divergence** when activated at epoch 5. Both temperatures (T=1.0 and T=0.5) show the same pattern:

1. **Epochs 1-4** (uniform sampling warmup): Normal convergence, metrics improving steadily
2. **Epoch 5**: Error maps computed from EMA model, last epoch with uniform sampling — still improving
3. **Epoch 6**: First epoch using error-weighted sampling — **immediate divergence** (loss 5-7x, grad_norm 4-8x, val metric regresses by 16-38 percentage points)

**Root cause**: Error-weighted sampling fundamentally shifts the training distribution. High-error points (wake regions, trailing edges, separation zones) get massively oversampled while smooth-flow regions (majority of the mesh) get undersampled. The model quickly "forgets" smooth regions, causing gradient explosion and loss spikes.

Even with log-transformed weights (power-law sampling: `probs ∝ errors^(1/T)` instead of the initial exponential `probs ∝ exp(errors/T)` which caused instant NaN), the distribution shift is too severe for stable training.

**Additional issues encountered** (all fixed in code):
- bf16 autocast in error map computation → CUBLAS_STATUS_INTERNAL_ERROR on variable-size einsum in Transolver slice attention. Fixed: fp32 inference for error maps.
- Missing `torch.no_grad()` → gradient accumulation OOM across ~2400 forward passes
- fp32 error map computation over 394 cases × ~300k points is extremely slow (~30-60 min per epoch), consuming the majority of the 180-min training budget

### Suggested follow-ups

1. **Error-weighted loss weighting (not sampling)**: Keep uniform sampling but weight the loss by per-point error magnitude. This preserves the full distribution while focusing gradient signal on hard regions.
2. **Gradual mixing**: Ramp from 100% uniform to 80/20 uniform/error-weighted over many epochs instead of switching abruptly at epoch 5.
3. **Error-stratified sampling**: Divide points into quantiles by error, ensure each quantile is proportionally represented (anti-collapse guarantee).
4. **Consider this hypothesis dead for direct surface point sampling** — the distribution shift problem is fundamental to importance sampling in point cloud regression.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3243 because ADVISOR: Closing — dead end. Both trials crashed at epoch 6 (first activation of error-weighted sampling). T=1.0: 40.79%, T=0.5: 39.31%. Classic importance-sampling distribution-shift collapse: hard points get massively overweighted, smooth regions starve, gradient norms explode 15-29x. Additional blockers: fp32 error-map computation costs 30-60 min/epoch (throughput-prohibitive), and weighted sampling creates train/eval mismatch (same fundamental issue as #3259 curvature loss). Thorough diagnosis from the student. Note: the student's suggestion of error-weighted LOSS (not sampling) avoids the distribution shift and throughput issues, but would still face the train/eval mismatch concern..
